pacer/pacer.go

```diff
@@ -1,25 +1,39 @@
-// pacer is a utility package to make pacing and retrying API calls easy
+// Package pacer makes pacing and retrying API calls easy
 package pacer
 
 import (
+	"math/rand"
 	"sync"
 	"time"
 
 	"github.com/ncw/rclone/fs"
 )
 
+// Pacer state
 type Pacer struct {
-	minSleep       time.Duration // minimum sleep time
-	maxSleep       time.Duration // maximum sleep time
-	decayConstant  uint          // decay constant
-	pacer          chan struct{} // To pace the operations
-	sleepTime      time.Duration // Time to sleep for each transaction
-	retries        int           // Max number of retries
-	mu             sync.Mutex    // Protecting read/writes
-	maxConnections int           // Maximum number of concurrent connections
-	connTokens     chan struct{} // Connection tokens
+	mu                 sync.Mutex    // Protecting read/writes
+	minSleep           time.Duration // minimum sleep time
+	maxSleep           time.Duration // maximum sleep time
+	decayConstant      uint          // decay constant
+	pacer              chan struct{} // To pace the operations
+	sleepTime          time.Duration // Time to sleep for each transaction
+	retries            int           // Max number of retries
+	maxConnections     int           // Maximum number of concurrent connections
+	connTokens         chan struct{} // Connection tokens
+	_calculatePace     func(bool)    // switchable pacing algorithm
```

Comment:
I haven't seen that before. Since it is unexported and not in a package, I would just add a comment along the lines of "This function assumes that the 'mu' mutex lock is held by the caller", or "This function may not be called concurrently".